gobblin-hive-registration/src/main/java/gobblin/hive/HiveRegister.java

```diff
@@ -0,0 +1,302 @@
+import gobblin.hive.spec.activity.Activity;
+import gobblin.util.ExecutorsUtils;
+import gobblin.util.executors.ScalingThreadPoolExecutor;
+import lombok.extern.slf4j.Slf4j;
+
+
+/**
+ * A class for registering Hive tables or partitions.
+ *
+ * <p>
+ * An instance of this class is initialized with a {@link State} object. To register a table or partition,
+ * properties {@link #HIVE_DB_ROOT_DIR}, {@link #HIVE_OWNER} are required. They are optional for de-registering
+ * or checking the existence of a table or partition.
+ * </p>
+ * <p>
+ * The {@link #register(HiveSpec)} method is asynchronous and returns immediately. Registration is performed in a
+ * thread pool whose size is controlled by {@link #HIVE_REGISTER_THREADS}
+ * </p>
+ *
+ * @author ziliu
+ */
+@Slf4j
+@Alpha
+public class HiveRegister implements Closeable {
+
+  public static final String HIVE_DB_ROOT_DIR = "hive.db.root.dir";
+  public static final String HIVE_OWNER = "hive.owner";
+  public static final String HIVE_REGISTER_THREADS = "hive.register.threads";
+  public static final int DEFAULT_HIVE_REGISTER_THREADS = 20;
+
+  private static final String HIVE_DB_EXTENSION = ".db";
+
+  private final HiveRegProps props;
+  private final Optional<String> hiveDbRootDir;
+  private final Optional<String> hiveOwner;
+  private final HiveMetaStoreClient client;
+
+  private final ListeningExecutorService executor;
+  private final List<Future<Void>> futures = Lists.newArrayList();
+  private final Striped<Lock> locks = Striped.lazyWeakLock(Integer.MAX_VALUE);
+
+  public HiveRegister(State state) throws MetaException {
+    this.props = new HiveRegProps(state);
+    this.hiveDbRootDir = this.props.getDbRootDir();
+    this.hiveOwner = this.props.getHiveOwner();
+    this.client = new HiveMetaStoreClient(new HiveConf());
+    this.executor = MoreExecutors.listeningDecorator(
+        ScalingThreadPoolExecutor.newScalingThreadPool(0, this.props.getNumThreads(), TimeUnit.SECONDS.toMillis(10),
+            ExecutorsUtils.newThreadFactory(Optional.of(log), Optional.of(getClass().getSimpleName()))));
+  }
+
+  /**
+   * Register a table or partition given a {@link HiveSpec}.
+   *
+   * This method is asynchronous and returns immediately.
+   *
+   * @return a {@link ListenableFuture} for the process of registering the given {@link HiveSpec}.
+   */
+  public ListenableFuture<Void> register(final HiveSpec spec) throws IOException {
+    ListenableFuture<Void> future = this.executor.submit(new Callable<Void>() {
+
+      @Override
+      public Void call() throws Exception {
+
+        if (spec instanceof HiveSpecWithPredicates && !evaluatePredicates((HiveSpecWithPredicates) spec)) {
+          log.info("Skipping " + spec + " since predicates return false");
+        }
+
+        if (spec instanceof HiveSpecWithPreActivities) {
+          for (Activity<HiveMetaStoreClient> activity : ((HiveSpecWithPreActivities) spec).getActivities()) {
+            activity.execute(HiveRegister.this.client);
```

Comment:
Not really. This is for the use case of dropping the hourly partitions after registering the daily partition. I'm not even sure whether dropping multiple partitions from the same table can be parallelized or not, but even if it can it likely won't bring any significant performance gain.